clients/src/main/java/org/apache/kafka/clients/consumer/internals/PartitionAssignor.java

```diff
@@ -51,14 +51,15 @@
      * {@link #assign(Cluster, Map)}.
      * @param topics Topics subscribed to through {@link org.apache.kafka.clients.consumer.KafkaConsumer#subscribe(java.util.Collection)}
      *               and variants
+     * @param ownedPartitions The partitions currently assigned to this consumer
      * @return Non-null subscription with optional user data
      */
-    Subscription subscription(Set<String> topics);
+    Subscription subscription(Set<String> topics, List<TopicPartition> ownedPartitions);
```

Comment:
Should we add a new default method to avoid breaking existing `PartitionAssignor` implementations? 

To be honest, I find this API a bit backwards. Most of the state is actually provided by the consumer. The only thing the assignor provides is the user data. I wonder if we should try to have an API like this:

```java
class AssignorSubscriptionData {
  ByteBuffer userData;
}

AssignorUserData subscriptionData(Subscription subscription);
```

Or something like that.

Reply:
By the way, once we figure out what this API should look like, we need to document the changes in KIP-429.

Reply:
I see what you're saying, since there is clear ownership of Subscription fields:  `PartitionAssignor` can inject some userData (maybe better named assignorData?) but `ConsumerCoordinator` sets all the other fields. So maybe it should just call some `PartitionAssignor#userData` and build the subscription itself. 

This seems cleaner, but if we do this and someone actually does implement a custom Assignor that 
1) for some reason needs to know TopicList, we irreversibly break it (this seems unlikely)
2) injects their own userData, either we:
2a. break their implementation by removing the `PartitionAssignor#subscription` 
2b. ignore their userData by deprecating `PartitionAssignor#subscription` and hope they read the docs/logs to discover they need to implement a new method 
2c. implement a really awkward temporary upgrade path, maybe by deprecating `PartitionAssignor#subscription` but continuing to call it, and use that userData if non-empty and otherwise use the userData from `PartitionAssignor#userData`

Reply:
Of these, I actually think 2a is pretty reasonable --- it is in the internal package (and the error/fix is quite clear). But if we don't want to break *anything*  we should just go with the current implementation of this PR (but add back the default) so that `ConsumerCoordinator` either passes in both fields or neither. Can always do the refactoring/cleanup for 3.0

WDYT @hachikuji @guozhangwang

Reply:
Ugh. I hadn't realized that `PartitionAssignor` was put in `consumer.internals`. That is really a bug since it is definitely a public API. But possibly it suggests an opportunity to kill two birds with one stone. Rather than attempting a fine-grained deprecation strategy, what if we deprecate the entire class? We can start a new API in the `consumer` package and move all of the KIP-429 changes there. Thoughts?

Reply:
Yeah it should not be `internals` at the very beginning.

About @ableegoldman 's worry that if user do not read the docs carefully and did not implement the new interface, by checking the code I think it should be okay: note that the config `PARTITION_ASSIGNMENT_STRATEGY_CONFIG` is only a type of `LIST` of class names in `PartitionAssignor.class`, which means we can actually enforce that these classes to be from the new interface programmatically since if users forget to do so, they will get a runtime ClassCast from `objects.add(t.cast(o));` in `getConfiguredInstances`.

But bear in mind that this also means this change would doom to require user's code changes, and as a result there's no point deprecating the old interface but just remove it directly.

Given there maybe about 10 people in the world who would be seriously impacted, I feel okay to just make this breaking change and require user code changes.